src/apps/external_executor/external_executor.cpp

```diff
@@ -641,19 +701,117 @@ namespace externalexecutor
               break;
             }
           }
+
+          out_stream->stream_msg(result);
         }
 
-        return ccf::grpc::make_success(results);
+        return ccf::grpc::make_success();
       };
 
       make_command_endpoint(
         "/temp.Test/RunOps",
         HTTP_POST,
-        ccf::grpc_command_adapter<
+        ccf::grpc_command_unary_stream_adapter<
           std::vector<temp::OpIn>,
-          std::vector<temp::OpOut>>(run_string_ops),
+          temp::OpOut>(run_string_ops),
         ccf::no_auth_required)
         .install();
+
+      auto sub =
+        [this](
+          ccf::endpoints::CommandEndpointContext& ctx,
+          externalexecutor::protobuf::KVKey&& payload,
+          ccf::grpc::StreamPtr<externalexecutor::protobuf::KVKeyValue>&&
+            out_stream) {
+          externalexecutor::protobuf::KVKey key;
+          key.set_table(payload.table());
+          key.set_key(payload.key());
+
+          std::unique_lock<ccf::pal::Mutex> guard(subscribed_keys_mutex);
+
+          subscribed_keys.emplace(std::make_pair(
+            key.SerializeAsString(),
+            ccf::grpc::detach_stream(std::move(out_stream))));
+          LOG_INFO_FMT(
+            "Subscribed to updates for key {}", key.SerializeAsString());
+
+          return ccf::grpc::make_pending();
+        };
+      make_endpoint(
+        "/temp.Test/Sub",
+        HTTP_POST,
+        ccf::grpc_command_unary_stream_adapter<
+          externalexecutor::protobuf::KVKey,
+          externalexecutor::protobuf::KVKeyValue>(sub),
+        {ccf::no_auth_required})
+        .install();
+
+      auto pub = [this](
+                   ccf::endpoints::EndpointContext& ctx,
+                   externalexecutor::protobuf::KVKeyValue&& payload)
+        -> ccf::grpc::GrpcAdapterResponse<google::protobuf::Empty> {
+        externalexecutor::protobuf::KVKey key;
+        key.set_table(payload.table());
+        key.set_key(payload.key());
+
+        std::unique_lock<ccf::pal::Mutex> guard(subscribed_keys_mutex);
+
+        auto search = subscribed_keys.find(key.SerializeAsString());
+        if (search != subscribed_keys.end())
+        {
+          LOG_TRACE_FMT(
+            "Publishing update for {}:{}", payload.table(), payload.key());
+          if (!search->second->stream_msg(payload))
+          {
+            // Manual cleanup of closed streams. We should have a close
+            // callback for detached streams to cleanup resources when
+            // required instead
+            subscribed_keys.erase(search);
+          }
+        }
+        else
+        {
+          return ccf::grpc::make_error(
+            GRPC_STATUS_NOT_FOUND,
+            fmt::format(
+              "Updates for key {} in table {} has no subscriber",
+              payload.key(),
+              payload.table()));
+        }
+
+        return ccf::grpc::make_success();
+      };
```

Comment:
This sample isn't actually interacting with the KV at all, so its odd that it uses `KVKeyValue`. I think it's a dangerous sample, because it _looks_ like a event-based notification of KV modifications, but this is executing entirely before commit/consensus.

I think these Pub/Sub endpoints should use new message types, with a simpler key indexing into the `subscribed_keys` table. If we want to build a sample of pub-sub event notification of changes to the KV, I think we need to do that in an indexer (and think about/comment on the reliability guarantees).

Reply:
Done.